### PR TITLE
Use TypeScript

### LINKED ISSUE #8: Migrate to TypeScript

### PR DESCRIPTION
**Pull request by braxtonhall**
resolves #8

### CODE REVIEW

File: extension/ts/copy.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	insertSaveData(parent, saveData);
+};
+
+const getSaveData = (parent: HTMLElement) => {
+	const elements = getElementsByTags(parent, RELEVANT_TAGS);
+	return elements.reduce((saveData: SaveData, element: any) => {
+		// We can't change hidden elements because LibraryThing relies
+		// on hidden form inputs to send additional, form-specific metadata
+		// on save
+		if (element && element.id && element.type !== "hidden") {
```
Comment (adirar111) [inline extension/ts/copy.ts:26]: super nit: i think `element && element.id` can be made shorter by doing `!!element?.id`

File: extension/ts/copy.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
+import { RELEVANT_TAGS } from "./constants";
+import { SaveData } from "./types";
```
Comment (adirar111) [inline extension/ts/copy.ts:2]: created a type for this

File: extension/ts/copy.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		// We can't change hidden elements because LibraryThing relies
+		// on hidden form inputs to send additional, form-specific metadata
+		// on save
+		if (element && element.id && element.type !== "hidden") {
+			let saveElement = element;
+			if (element.id.startsWith(COLLECTIONS_ID_PREFIX)) {
+				const span =
+					element.parentElement.getElementsByTagName("span")[0];
+				saveElement =
+					saveData[COLLECTIONS_KEY][span.textContent] || element;
```
Comment (adirar111) [inline extension/ts/copy.ts:54]: super nit: this probably always exists, but for the back burner maybe can change to `element.parentElement.getElementsByTagName("span")?.[0]` and `span?.textContent`

File: extension/ts/copy.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	const td = document.createElement("td");
+	td.appendChild(button);
+	element.appendChild(td);
+};
+
+const appendRow = (editForm: HTMLElement) => (table: HTMLTableElement) => {
+	const row = document.createElement("tr");
+	appendButton(row, "SAVE", saveData(editForm));
+	appendButton(row, "LOAD", loadData(editForm));
+	const [body] = Array.from(table.getElementsByTagName("tbody"));
```
Comment (adirar111) [inline extension/ts/copy.ts:81]: i added `Array.from` cuz ts error

File: package.json [hunk 5]
```diff
@@ -2,5 +2,19 @@
...
+		"ts-loader": "^9.4.1",
+		"typescript": "^4.8.4",
+		"webpack": "^5.74.0",
+		"webpack-cli": "^4.10.0",
+		"prettier": "^2.7.1"
+	},
+	"scripts": {
+		"watch": "webpack --watch --mode development",
+		"build": "webpack --mode production",
+		"pretty": "yarn prettier --editorconfig true --write extension/ts/*.ts"
```
Comment (adirar111) [inline package.json:18]: added prettier

**Review by adirar111 — APPROVED** [review]
ts woooooo

left a few tiny nits and went through and added prettier + some missing types that ts was warning about

File: extension/ts/copy.ts (hunk 3 above)
Comment (braxtonhall) [reply extension/ts/copy.ts:54]: `getElementsByTagName` always will give you back a collection, but `span` could be `undefined`. So yeah you'd want the optional chaining on `span.textContent`. I think that should be in another PR for generally making the code safer.

File: extension/ts/copy.ts (hunk 1 above)
Comment (braxtonhall) [reply extension/ts/copy.ts:26]: `!!` not needed.
```typescript
if (element?.id && element.type !== "hidden") {
```

Comment (braxtonhall) [reply extension/ts/copy.ts:26]: I've been thinking about this, and I think I prefer it with the explicit `element`.

If it were
```typescript
if (element?.id) {
```
I'd say that that's _perfect_. 

Reasoning:

`element?.id` accomplishes two checks:
1. `element` is truthy
1. `id` is truthy

However, `element.type` depends on the check "`element` is truthy", which is hidden inside the `element?.id`, and contains a check that is unrelated to `element.type`: "`id` is truthy".

You could put the "`element` is truthy" inside both expressions, however `element?.type !== "hidden"` on its own would be a bug, so I don't think that's the right way.

Comment (braxtonhall) [reply extension/ts/copy.ts:26]: (I often think shorter is bad for readability and maintainability)

**Pull request merged**